### ISSUE TITLE
How add 4 classes to trained(exist) weights?

### ISSUE THREAD

**Issue by RushNuts** [AlexeyAB/darknet#459]
Hello! How i can add 4 classes to exist weights? In my case i have 4 my classes and i want add them to yolo-voc.weights. How i can do this? Thanks

I tried add my .txt&.jpg to obj folder, create test, train; copy yolo-voc.cfg and change classes to 21, filtres to 135, after command: ./darknet partial cfg/yolo-voc-21.cfg yolo-voc.weights yolo-voc.conv.23 23
unfortunately nothing succeeded

**Comment by AlexeyAB** [comment]
No. You should train from the beginning with all 24 classes in your training dataset.

Also if you want to train 24 classes, then filters=145
`(24+5)*5 = 29*5 = 145`

**Comment by RushNuts** [comment]
Okay, thanks. Then, i have next question.

Now i have VOCdevkit folder with next files: 
VOCtest_06-Nov-2007.tar
VOCtrainval_06-Nov-2007.tar
VOCtrainval_11-May-2012.tar

Unzip them. I'm add my .jpg files to VOCdevkit>VOC2007>JPEGImages, and .txt files to VOCdevkit>VOC2007>labels. When i run voc_label.py script, but the files I added did not appear in any of the .txt files.

So, how do I properly add my classes to the VOC? Thanks for your answer!

**Comment by AlexeyAB** [comment]
1. run `voc_label.py` you will get: 2007_test.txt, 2007_train.txt, 2007_val.txt, 2012_train.txt, 2012_val.txt
2. run Yolo_mark to get list of jpg-files in the file `train_obj.txt` with full paths in a such way: `yolo_mark.exe D:\Darknet2\Yolo_mark\x64\Release\data\img data/train_obj.txt data/obj.names`  https://github.com/AlexeyAB/Yolo_mark
3. Copy this `train_obj.txt` to the same directory where `2007_test.txt`
4. Do `type train_obj.txt 2007_train.txt 2007_val.txt 2012_*.txt > train.txt`
5. Train your model with this `train.txt` in the `obj.data` file and changed `classes=24 filter=145` in the yolo-voc.2.0.cfg file:
`darknet.exe detector train data/obj.data yolo-voc.2.0.cfg darknet19_448.conv.23`

**Comment by RushNuts** [comment]
Sorry the dumb question, train_obj.txt is it a file with my image classes?

**Comment by RushNuts** [comment]
AlexeyAB, could you answer a few more questions, please?
1. How can I leave only the car classes and the person from the VOC? In order to add to these two classes 4 classes from the my dataset. Is this possible?
2. Can I learn the same way with the help of COCO data?
3.Why do you need the team partial? darknet.exe partial cfg / tiny-yolo-voc.cfg tiny-yolo-voc.weights tiny-yolo-voc.conv.13 13

Thank you for your work!

**Comment by lesreaper** [comment]
So Alex, are you saying you add the custom images to the VOC images path, along with adding the custom training.txt and test.txt contents into the respective VOC versions of those files, and then train with the entire VOC dataset?

**Comment by AlexeyAB** [comment]
@RushNuts 

> Sorry the dumb question, train_obj.txt is it a file with my image classes?

Yes. 

1. Leave here only `car` and `person`: https://github.com/AlexeyAB/darknet/blob/df076653e00db69c6fb57869981f5196a8f55e70/scripts/voc_label.py#L9
    So if there are 2 classes from voc and 4 your, than `filter = 55`

2. Try to use this manual for COCO training: https://pjreddie.com/darknet/yolo/#train-coco
    And https://groups.google.com/forum/#!topic/darknet/jpQHwTiC4go

3. By this way you can get pre-trained weights-file with weights only from the first 13 layers which are the same in the `tiny-yolo-voc.cfg` and in your `tiny-yolo-obj.cfg` that based on tiny-yolo-voc.cfg. Last convolutional layer is not the same in voc and your cfg files, because you change `filter=55`

**Comment by AlexeyAB** [comment]
@lesreaper 

> are you saying you add the custom images to the VOC images path, along with adding the custom training.txt and test.txt contents into the respective VOC versions of those files, and then train with the entire VOC dataset?

For example, VOC dataset can be in the directory `E:\VOC2007_2012\VOCdevkit`
But your dataset can be in the directory `D:\Darknet2\Yolo_mark\x64\Release\data\img`

So result `train.txt` will contain full paths to each image
```
E:\VOC2007_2012\VOCdevkit\VOC2007\JPEGImages\000001.jpg
D:\Darknet2\Yolo_mark\x64\Release\data\img\air1.jpg
```
...

**Comment by RushNuts** [comment]
@AlexeyAB Thank you very much! Will try tomorrow. 

I'm work on Ubuntu, Yolo_mark show cuda version error (now installed cuda 8.0)

**Comment by RushNuts** [comment]
@AlexeyAB Thank you very much! Will try tomorrow.

I'm work on Ubuntu, Yolo_mark show cuda version error (now installed cuda 8.0)

I forget ask. If train_obj.txt is contain paths for train my classes, how I create test.txt for them?

**Comment by AlexeyAB** [comment]
@RushNuts 
Do in the command line (on Windows or Linux):
`type train_obj.txt 2007_train.txt 2007_val.txt 2012_*.txt > train.txt`

**Comment by RushNuts** [comment]
@AlexeyAB
sorry, I incorrectly asked a question. for each class there are two files (one for training, another for testing). in this case, train_obj.txt comes out that contains only for learning by my classes. and test_obj.txt will not exist for my classes? sorry, I'm completely confused now trying to combine VOC classes with my classes

**Comment by RushNuts** [comment]
@AlexeyAB 

Did as you wrote yesterday. Faced a problem, now I am voc_label.py changed to person and car. I started the script.

How can I create voc.names? Since the car is 6 class, people is 14 class.

I'm create obj.data.
classes = 4
train = train.txt
valid = test.txt
names = data / voc.names
backup = backup /

wok.names:
aeroplane
bicycle
bird
boat
bottle
bus
car - I chose
cat
chair
cow
diningtable
dog
horse
motorbike
person - I chose
pottedplant
sheep
sofa
train
tvmonitor
no-parking

My should be the following ?:
car
person
myclass
myclass

**Comment by AlexeyAB** [comment]
> How can I create voc.names? Since the car is 6 class, people is 14 class.

Look at the voc txt labels files `VOCdevkit\VOC2007\labels`, there should be 0 and 1 classes id in the txt-files.

**Comment by RushNuts** [comment]
Yes, I should in every .txt file change class id? 
6 to 0
14 to 1

**Comment by AlexeyAB** [comment]
@RushNuts No, there are already 0 and 1 :) Just look at these files.

voc.names should be:
```
car
person
myclass_1
myclass_2
myclass_3
myclass_4
```

**Comment by RushNuts** [comment]
Oh, sorry for stupid questions.thanks

One more thing. Above, I asked about files test.txt and train_obj.txt. Can you explain ?

"for each class there are two files (one for training, another for testing). in this case, train_obj.txt comes out that contains only for learning by my classes. and test_obj.txt will not exist for my classes? sorry, I'm completely confused now trying to combine VOC classes with my classes"

**Comment by AlexeyAB** [comment]
@RushNuts In your case I think you can just fuse all _train, _val, _test.txt files into one, and train using it: `type train_obj.txt 2007_train.txt 2007_val.txt 2012_*.txt > train.txt`


> for each class there are two files (one for training, another for testing). in this case, train_obj.txt comes out that contains only for learning by my classes. and test_obj.txt will not exist for my classes? sorry, I'm completely confused now trying to combine VOC classes with my classes"

Not for each class. But for each dataset. PascalVOC-2007 has (test, val, train) and PascalVOC-2012 has (val, train, test):
* train - for training
* val - for checking when you should stop training
* test - for checking detection accuracy

It is required to check which model is the best. But if you want to get the best accuracy - you should train using all datasets (test, val, train).

**Comment by RushNuts** [comment]
@AlexeyAB 
I understand correctly, in my case(my classes from yolo_mark) there will be only train_obj.txt and that's all?

**Comment by AlexeyAB** [comment]
@RushNuts Yes.

**Comment by RushNuts** [comment]
Thank you very much! I hope I did not tire much :D

**Comment by RushNuts** [comment]
@AlexeyAB I forget ask you. how to change, so that weights are saved every 1000 iterations, not 100?

**Comment by TheMikeyR** [comment]
@RushNuts in `src/detector.c` change the number in this line
https://github.com/AlexeyAB/darknet/blob/164ba34279a4d486d6e49409de791817b62ad1ac/src/detector.c#L182
to:

`if(i >= (iter_save + 1000)) {`

**Comment by RushNuts** [comment]
@TheMikeyR Okay, thanks! I will try

**Comment by RushNuts** [comment]
@AlexeyAB
Hello! I tried train, with 2 classes from VOC and my classes. 
Unfortunately detection results is bad. 

Maybe something wrong with id? 
You says car and person already 0 and 1 id. So, I change my classes id;
Myclasses id - 2
Myclasses id- 3

**Comment by AlexeyAB** [comment]
@RushNuts 

* So did you use 2 VOC classes and 2 your classes?
* Can you show images with bad detections?
* How many iterations did you train, what is average loss and what is mAP and IoU using `./darknet detector map obj.data obj.cfg obj.weights`?

**Comment by RushNuts** [comment]
@AlexeyAB 

- Now i use 3 classes from VOC and 4 my classes. change classes in obj.data to 7. yolo-voc.2.0.cfg to 7 and filtres 60. 
voc.names:
bus
car
person
myclass
myclass
myclass
myclass
- Yes. My class is detect, but car no detect.
![predictions](https://user-images.githubusercontent.com/11484847/37563528-a1459874-2a93-11e8-952f-c6d26910a132.jpg)

- 10000 iterations. After 10000 Ceased to detect at all and does not detect anything. run map command:
detections_count = 134682, unique_truth_count = 5942  
class_id = 0, name = bus, 134682 ap = 63.88 % 
class_id = 1, name = car, 	 ap = 63.93 % 
class_id = 2, name = person, 	 ap = 60.51 % 
class_id = 3, name = speed, 	 ap = 0.00 % 
class_id = 4, name = stop, 	 ap = 0.00 % 
class_id = 5, name = roadworks, 	 ap = 0.00 % 
class_id = 6, name = zebra, 	 ap = 0.00 % 
 for thresh = 0.24, precision = 0.63, recall = 0.67, F1-score = 0.65 
 for thresh = 0.24, TP = 3979, FP = 2293, FN = 1963, average IoU = 44.21 % 

 mean average precision (mAP) = 0.269036, or 26.90 % 
Total Detection Time: 97.000000 Seconds

**Comment by AlexeyAB** [comment]
@RushNuts 

* As I see you classes aren't detected. Only VOC classes are detected.
* Check that paths to your images there are in the `train.txt` that you use in the `obj.data`
* Try to use Yolo_mark to check your dataset: https://github.com/AlexeyAB/Yolo_mark
    `./yolo_mark /your_images_folder dont_use_train.txt /data/obj.names` - can you see correct bounded boxes?

**Comment by RushNuts** [comment]
@AlexeyAB 
- Basically, my classes are detected, but the VOC classes are not always there.
- Paths to check, everything is correct.

**Comment by AlexeyAB** [comment]
what classes= and filters= do you use in the cfg-file?
and show you obj.data file

> Basically, my classes are detected, but the VOC classes are not always there.

As I see precision for VOC classes is good, but precision for your classes is zero.

> class_id = 0, name = bus, 134682 ap = 63.88 %
> class_id = 1, name = car, ap = 63.93 %
> class_id = 2, name = person, ap = 60.51 %
> class_id = 3, name = speed, ap = 0.00 %
> class_id = 4, name = myclass2, ap = 0.00 %
> class_id = 5, name = myclass3, ap = 0.00 %
> class_id = 6, name = myclass4, ap = 0.00 %

**Comment by RushNuts** [comment]
@AlexeyAB 
use 7 classes, filtres = 60 in yolo-voc.2.0.cfg file, which i use.
obj.data:
classes = 7
train = train.txt
valid = test.txt
names = data/voc.names
backup = backup/

**Comment by AlexeyAB** [comment]
Check, is test.txt the same as train.txt?

**Comment by RushNuts** [comment]
@AlexeyAB

checked, different, no repetition in the files. In the train added more ways to my files for my classes, in the test only VOC

**Comment by AlexeyAB** [comment]
@RushNuts Copy train.txt to test.txt so it will be the same.
And then do `./darknet detector map`... again.

**Comment by RushNuts** [comment]
@AlexeyAB 
do it. train copy into test, results:

class_id = 0, name = bus, 495041 ap = 74.98 % 
class_id = 1, name = car, 	 ap = 64.48 % 
class_id = 2, name = person, 	 ap = 68.69 % 
class_id = 3, name = speed, 	 ap = 58.55 % 
class_id = 4, name = myclass2, 	 ap = 52.03 % 
class_id = 5, name = myclass3, 	 ap = 85.83 % 
class_id = 6, name = myclass4, 	 ap = 43.20 % 
 for thresh = 0.24, precision = 0.67, recall = 0.69, F1-score = 0.68 
 for thresh = 0.24, TP = 20130, FP = 9866, FN = 9118, average IoU = 47.83 % 

 mean average precision (mAP) = 0.639660, or 63.97 % 
Total Detection Time: 289.000000 Seconds

**Comment by AlexeyAB** [comment]
So, not bad results. A little bit problems only in myclass4 and myclass2. If these classes are small, then try to increase resolution `width=608 heigh=608` in cfg and do `./darknet detector map` again.

----

Or you can try 
* to increase resolution in the cfg width=608 heigh=608
* set random=1 in cfg
* do `./darknet detector calc_anchors data/voc.data -num_of_clusters 5 -final_width 19 -final_heigh 19 -show` and then set `anchors=` in your cfg-file from the generated `anchors.txt` file
* train again about 45000 iterations

**Comment by RushNuts** [comment]
@AlexeyAB 
Well, why then when I check on the video or on the images, he does not see the car or people?

**Comment by AlexeyAB** [comment]
@RushNuts 

Check that in your (not VOC) dataset there aren't non-labeled objects from VOC-classes: buses, cars, persons.

Try to train from the begining with `width=608 height=608 random=1` and new anchors.

About anchors you can read: https://github.com/AlexeyAB/darknet/issues/418

**Comment by RushNuts** [comment]
@AlexeyAB Do not quite understand. My data set does not contain VOC data

**Comment by AlexeyAB** [comment]
For example, you have your images with signs Speed, but these signs are on the road where can be cars. 
If these cars are not labeled in your images - this is very bad for training.

**Comment by RushNuts** [comment]
@AlexeyAB Yes, there are images where only the sign is marked. Because of what so? With its data set, there were no problems until I tried with VOC.

**Comment by AlexeyAB** [comment]
If the neural network sees an object that is not labeled, then it will learn to never detect it.

**Comment by RushNuts** [comment]
Even if this object is marked in another class?

**Comment by AlexeyAB** [comment]
Even if this object is marked in another images.

**Comment by RushNuts** [comment]
I understand correctly? If in my data set on the image only the sign of speed is marked (even if there is a person or car besides the sign), does this have a bad effect on training? However, this image is not present in the VOC data set

**Comment by AlexeyAB** [comment]
Yes

**Comment by RushNuts** [comment]
Will try mark them, so very thank you

**Comment by RushNuts** [comment]
@AlexeyAB I tried. This did not solve the problem, the detection became worse. Can there be a problem in that my images have a high resolution (1366x766 approximately)

**Comment by lesreaper** [comment]
@AlexeyAB I'm wondering if this might be part of my issue as well, similar to @RushNuts .

I get 1.000 mAP, tests all work, but I have terrible production use.

I'm checking people coming in and out of doors. I have thousands of images of enter and exit, as well as two custom people.

It sounds like I have to label every person in these images if I'm going to train a custom dataset using the pre-trained Tiny-Yolo features.

What if I'm also trying to train the system to recognize two specific persons, for which I have 2 classes, each with 2k images. 

How do I keep the system from getting confused if it is a person or a specific person?

**Comment by AlexeyAB** [comment]
@RushNuts Hi, 

* what mAP can you get now?
* and what `width height random` params do you use?
* did you recalculate anchors for your dataset?

To increase precision you can
* just increase `width height` to any value multiple of 32, and do detection without re-training
* increase `width height` to any value multiple of 32, set `random=1`, and training again
* use models renset_yolo or denset_yolo, change `classes, filters, num, anchors, width, height, random` and train:
    * https://github.com/AlexeyAB/darknet/blob/master/build/darknet/x64/resnet152_yolo.cfg
    * https://github.com/AlexeyAB/darknet/blob/master/build/darknet/x64/densenet201_yolo.cfg

**Comment by RushNuts** [comment]
@AlexeyAB 
- mAP is:
bus = 78.655
car = 80.57%
person = 74.59%
myclass1=0%
myclass2=0%
myclass3=0%
myclass4=0%

mAP = 33.40%

-I'm use standart 416x416 and 608x608 (without change anchors and randon value to 1)
-i try change 608x608, random=1 and new anchors. When i start train in some time i get CUDA error.

Now i change only widht and height to 608x608 and now start new train

**Comment by AlexeyAB** [comment]
@lesreaper 

> I get 1.000 mAP, tests all work, but I have terrible production use.

Do you get 100% mAP and terrible production? This is very strange. It can be only if your Training dataset is very different from the Test dataset.

----

> It sounds like I have to label every person in these images if I'm going to train a custom dataset using the pre-trained Tiny-Yolo conv.13 weights, even if I'm not adding "person" to my list of objects.

You should label **every person on every image** in two cases:
* if you want to detect persons
* if you want to detect very similar objects (person coming in door, person out of door) - in this case 
    * obj-0: person coming in door
    * obj-1: person out of the door
    * obj-2: any other person

It is independent on do you use pre-trained `tiny-yolo.conv.13.weights` or not.

----

> What if I'm also trying to train the system to recognize two specific persons, for which I have 2 classes, each with 2k images. Would I add "persons from VOC", and then add each of these individuals pictures as a separate class? Subclass?

So use:
* obj-0: required person-1
* obj-1: required person-2
* obj-2: any other person

**Comment by AlexeyAB** [comment]
@RushNuts 

> mAP is:
> bus = 78.655
> car = 80.57%
> person = 74.59%
> myclass1=0%
> myclass2=0%
> myclass3=0%
> myclass4=0%

So your training or validation dataset does not contain images of your objects (myclass1 - 4). 
**You should fix it.**

**Comment by RushNuts** [comment]
@AlexeyAB Yes, i fix it. Percents is high, above 85%. But when i start test on real video, my classes have is so bad detections :( . in .cfg change to 608x608, nothing is change. Now i start training with 608x608 value

**Comment by AlexeyAB** [comment]
> Yes, i fix it.

Can you show now mAP and AP for each class?

**Comment by RushNuts** [comment]
@AlexeyAB 

mAP is:
bus = 76.34
car = 82.53%
person = 72.44%
myclass1=80.65%
myclass2=88.21%
myclass3=79.32%
myclass4=76.46%

mAP = 69.65%

**Comment by AlexeyAB** [comment]
@RushNuts This is strange that mAP lower than any AP, because mAP is average value of APs.

So try to train 608x608 random=1 and increased subdivision.

**Comment by RushNuts** [comment]
@AlexeyAB 
Hello again. Can i get video stream from gopro?

**Comment by lesreaper** [comment]
I don't know where to post this, but Yolo V3 just came out today!! https://pjreddie.com/darknet/yolo/

**Comment by AlexeyAB** [comment]
@RushNuts Hi, I think no. Gopro can't be used as WebCam or as IP-cam.
But you can use SjCam 4000 as web-cam to use with Yolo. Or use your smartphone: https://github.com/AlexeyAB/darknet#for-using-network-video-camera-mjpeg-stream-with-any-android-smartphone

**Comment by RushNuts** [comment]
@AlexeyAB Okay, thank you!

Btw, i check my last train with 608x608 and random=1. Results:
bus= 73,19%
car=65,78%
person=70,61%
myclass_1=87.81%
myclass_2=77.00%
myclass_3=84.92%
myclass_4=89.96%

mAP=79.01%

Detection is bad, have a false detections and sometimes do not see objects for which it is trained. how many would not train with different parameters, the detection is still the same bad :(

**Comment by AlexeyAB** [comment]
79.01 % mAP (mean AP) doesn't look like a bad result. 
You can see precision of any other detection algorithms (neural and not) using limited number of images - they have mean AP less than 79% : http://host.robots.ox.ac.uk:8080/leaderboard/displaylb.php?challengeid=11&compid=3

Do you use tiny-yolo-voc.cfg or yolo-voc.2.0.cfg?
Now you can try to use yolov3.cfg on Linux: https://github.com/pjreddie/darknet
Or wait when I will port it to Windows.

Also you should collect much more images for training.

**Comment by RushNuts** [comment]
Hi @AlexeyAB 
Can you tell me about some commands?
- darknet.exe detector test data/voc.data yolo-voc.cfg yolo-voc.weights **-i 0** - what does mean -i and 0?
- how work command? -http_port 8090 

Thank you very much!

**Comment by RushNuts** [comment]
@AlexeyAB i have one more question. i check your script "net_cam_voc.sh", it is for work with net cam? And this script just for only lan working camera?

**Comment by AlexeyAB** [comment]
@RushNuts This script for any network camera that uses HTTP-mjpeg-stream or RTSP stream.

**Comment by RushNuts** [comment]
@AlexeyAB Okay. 

btw, i'm working on linux and i try connect smartphone like a web camera. Use your manual. I get error "could not connect to web camera". What's wrong? Use both applications from Google play

**Comment by AlexeyAB** [comment]
@RushNuts Probably between your smartphone and PC there is NAT or proxy, so your phone doesn't have own exclusive IP, so in this case PC and Smartphone should be in the same Local Network. For example connected to the same router (Eth, Wifi)

**Comment by RushNuts** [comment]
@AlexeyAB 
You gave me a great idea about the proxy! I will try something and answer

**Comment by RushNuts** [comment]
@AlexeyAB check it! unfortunatelly proxy is not my problem(
1. i install Smart Web cam on my phone
2. connected via usb
3. lauch smart web cam app on my phone and get ip address
4. when try connect, i see error "Couldn't connect to webcam". in VLC not working too. Uh. I try Ip cam app too. 

i'm working on linux, but windows have same problem

**Comment by AlexeyAB** [comment]
@RushNuts So you should find the URL that can work in VLC.
When your phone is connected via USB - it can be in different mode, and only one of them allows IP connections.

**Comment by Vaquitta** [comment]
Hi Alexey,

This issue chain is great source of information for troubleshooting training custom set. 

Had  a question though. 
Your comment "No. You should train from the beginning with all 24 classes in your training dataset."

Does this mean every time i have a single new class that needs to appended to existing set of classes , i have to start training from begining , with all the classes ?

**Comment by AlexeyAB** [comment]
@Vaquitta Yes, if you want good accuracy.

**Comment by Vaquitta** [comment]
Dear Alexey,

Appreciate the quick response. 

If trade-off on accuracy is acceptable , how do we go about this append option. All the documents hint/instruction are for full training, hence not sure about append option..

Basically i wanted to benchmark such an "Append" vs Full training ...

**Comment by AlexeyAB** [comment]
@Vaquitta 

1. Or you can just train VOC-model `yolo-voc.weights` on the new dataset with only 4 new classes (20, 21, 22, 23 in the txt-files).

2. Or you can train VOC-model `yolo-voc.weights` on the dataset that contains Pascal-VOC dataset (0-19 classes) and your dataset (20-23 classes)

3. And then compare mAP with model trained from the begining using Pascal-VOC dataset (0-19 classes) and your dataset (20-23 classes): https://github.com/AlexeyAB/darknet/tree/47c7af1cea5bbdedf1184963355e6418cb8b1b4f#how-to-train-pascal-voc-data

**Comment by Vaquitta** [comment]
Thanks Alexey. 

I am planning to use the Yolov3.weights and the coco dataset. 
In option 2 you meant the Yolo-voc .weights(Yolov3.weights in my case) or the base pre-trained model like the darknet53.conv.74, ideally that is what would be a training from scratch isn't it ?

**Comment by AlexeyAB** [comment]
@Vaquitta

1.  `yolov3.weights` and modified `yolov3.cfg` (increased classes and filters in 3 yolo-layers)
Train only using your dataset.

2. `yolov3.conv.105` and modified `yolov3.cfg` (increased classes and filters in 3 yolo-layers)
File `yolov3.conv.105` you can get by command:
`darknet.exe partial cfg/yolov3.cfg yolov3.weights yolov3.conv.105 105`
Train using MSCOCO + your dataset.

3. `darknet53.conv.74` and modified `yolov3.cfg` (increased classes and filters in 3 yolo-layers)
Train using MSCOCO + your dataset.

**Comment by Vaquitta** [comment]
Thanks Alexey for the clarification, the picture is getting much clearer 

So in all the 3 cases i would assume the 
classes = 83 [Assuming i have 3 additional classes and already 80 classes of MSCOCO]
Filters = 264 [(Number of classes + 3) * 5]
In all cases i woulse use 

./darknet detector train cfg/cocoCustom.data cfg/yolov3Custom.cfg <weight-file> [  yolov3.weights for option 1, yolov3.conv.105 for option 2 and darknet53.conv.74 for option 3]

Option 2 , i am assuming we are getting the weights other than the final detection layer and then we are trying to train the detector , am i correct..
Option 3, its pre-trained till 74 layers hence we are re-training part of classifier and detector

**Comment by AshleyRoth** [comment]
@AlexeyAB, Hi! can I add additional several classes to already trained weights? My weights, just learned. There was a need to add several more classes 
Thanks!
